### DISCUSSION TITLE
Clickhouse is very slow to obtain the array

### DISCUSSION THREAD

**Discussion by YH2628624388** [ClickHouse/ClickHouse#56361] [Q&A] [answered]
I have a table in clickhouse. 
The table structure is as follows

`CREATE TABLE data_feature_store.table18
(
    `md5` String,
    `col187` Nullable(String),
    `col188` Nullable(String),
    `col189` Nullable(String),
    `col190` Array(Float64),
    `col191` Array(Float64),
    `col192` Nullable(Int32),
    `col193` Nullable(Int8),
    `col194` Nullable(String),
    `col195` Nullable(Int32),
    `col196` Nullable(Int32),
    `col197` Nullable(Int32),
    `col198` Nullable(Int32),
    `col199` Nullable(Int32),
    `col200` Nullable(Int32),
    `col201` Nullable(Int32),
    `col202` Nullable(Int32),
    `col203` Nullable(Int32),
    `col204` Nullable(Int32),
    `col205` Nullable(String),
    `col206` Nullable(String),
    `col207` Nullable(String),
    `col208` Nullable(Int32),
    `col209` Nullable(Int32),
    `col210` Nullable(Int32),
    `col211` Nullable(Int32),
    `col212` Nullable(Int32),
    `col213` Nullable(Int8),
    `col214` Nullable(Float64),
    `col215` Nullable(Int8),
    `col216` Nullable(Float64)
)
ENGINE = ReplacingMergeTree
ORDER BY md5
SETTINGS index_granularity = 8192`

when i get col190,col191 data by md5
took 50 seconds
sql：select col190,col191 from table18 where md5 IN (1000 md5.....)

took 25 seconds
sql：select col190 from table18 where md5 IN (1000 md5.....)

took 1 seconds
sql：select col187 from table18 where md5 IN (1000 md5.....)

I want to ask if there is any good solution and optimization method. I want to know why this is.

**Comment by YH2628624388** [comment]
There are about 50 million pieces of data in this table

**Reply by den-crane** [reply, marked as answer]
How many rows do you expect in this table in the future? You may try `SETTINGS index_granularity = 256`.

I guess the problem is the size of Array columns

Check the column sizes using

```
SELECT
    database,
    table,
    column,
    formatReadableSize(sum(column_data_compressed_bytes) AS size) AS compressed,
    formatReadableSize(sum(column_data_uncompressed_bytes) AS usize) AS uncompressed,
    round(usize / size, 2) AS compr_ratio,
    sum(rows) rows_cnt,
    round(usize / rows_cnt, 2) avg_row_size
FROM system.parts_columns
WHERE (active = 1) AND (database LIKE '%') AND (table LIKE '%')
GROUP BY
    database,
    table,
    column
ORDER BY size DESC;
```

**Reply by YH2628624388** [reply]
This table is expected to maintain 60 million rows.


After I changed the sql query, I found that col190 and col191 took up a lot of storage space.I can try the settings SETTINGS index_granularity = 256. to see if that solves the problem

Do you have any other suggestions?

database───────────┬─table───┬─column─┬─compressed─┬─uncompressed─┬─compr_ratio─┬─rows_cnt─┬─avg_row_size─┐
│ data_feature_store │ table18 │ col191 │ 15.00 GiB  │ 18.03 GiB    │         1.2 │ 61249051 │       316.07 │
│ data_feature_store │ table18 │ col190 │ 14.92 GiB  │ 18.03 GiB    │        1.21 │ 61249051 │       316.07 │
│ data_feature_store │ table18 │ md5    │ 1.72 GiB   │ 1.88 GiB     │        1.09 │ 61249051 │           33 │
│ data_feature_store │ table18 │ col205 │ 193.57 MiB │ 1.04 GiB     │        5.51 │ 61249051 │        18.27 │
│ data_feature_store │ table18 │ col206 │ 146.31 MiB │ 444.33 MiB   │        3.04 │ 61249051 │         7.61 │
│ data_feature_store │ table18 │ col188 │ 140.23 MiB │ 526.14 MiB   │        3.75 │ 61249051 │         9.01 │
│ data_feature_store │ table18 │ col214 │ 134.19 MiB │ 525.68 MiB   │        3.92 │ 61249051 │            9 │
│ data_feature_store │ table18 │ col209 │ 130.88 MiB │ 292.04 MiB   │        2.23 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col187 │ 125.91 MiB │ 405.25 MiB   │        3.22 │ 61249051 │         6.94 │
│ data_feature_store │ table18 │ col210 │ 122.95 MiB │ 292.04 MiB   │        2.38 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col196 │ 121.73 MiB │ 292.04 MiB   │         2.4 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col216 │ 116.92 MiB │ 525.68 MiB   │         4.5 │ 61249051 │            9 │
│ data_feature_store │ table18 │ col207 │ 114.18 MiB │ 321.98 MiB   │        2.82 │ 61249051 │         5.51 │
│ data_feature_store │ table18 │ col192 │ 107.34 MiB │ 292.04 MiB   │        2.72 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col189 │ 105.96 MiB │ 457.33 MiB   │        4.32 │ 61249051 │         7.83 │
│ data_feature_store │ table18 │ col194 │ 89.78 MiB  │ 321.97 MiB   │        3.59 │ 61249051 │         5.51 │
│ data_feature_store │ table18 │ col211 │ 88.78 MiB  │ 292.04 MiB   │        3.29 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col199 │ 73.43 MiB  │ 292.04 MiB   │        3.98 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col193 │ 62.93 MiB  │ 116.82 MiB   │        1.86 │ 61249051 │            2 │
│ data_feature_store │ table18 │ col200 │ 61.69 MiB  │ 292.04 MiB   │        4.73 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col201 │ 58.87 MiB  │ 292.04 MiB   │        4.96 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col203 │ 48.74 MiB  │ 292.04 MiB   │        5.99 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col212 │ 45.94 MiB  │ 292.04 MiB   │        6.36 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col202 │ 44.86 MiB  │ 292.04 MiB   │        6.51 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col213 │ 34.09 MiB  │ 116.82 MiB   │        3.43 │ 61249051 │            2 │
│ data_feature_store │ table18 │ col215 │ 32.07 MiB  │ 116.82 MiB   │        3.64 │ 61249051 │            2 │
│ data_feature_store │ table18 │ col208 │ 31.11 MiB  │ 292.04 MiB   │        9.39 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col197 │ 29.24 MiB  │ 292.04 MiB   │        9.99 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col198 │ 27.63 MiB  │ 292.04 MiB   │       10.57 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col204 │ 20.32 MiB  │ 292.04 MiB   │       14.37 │ 61249051 │            5 │
│ data_feature_store │ table18 │ col195 │ 17.00 MiB  │ 292.04 MiB   │       17.18 │ 61249051 │            5 │

**Reply by den-crane** [reply]
```
│ data_feature_store │ table18 │ col191 │ 15.00 GiB │ 18.03 GiB │ 1.2 │ 61249051 │ 316.07 │
│ data_feature_store │ table18 │ col190 │ 14.92 GiB │ 18.03 GiB │ 1.21 │ 61249051 │ 316.07 │
```

It works as expected.

>Do you have any other suggestions? 

no.

**Reply by YH2628624388** [reply]
it works,thanks
